src/twisted/web/test/test_web.py

```diff
@@ -751,6 +751,41 @@ def test_retrieveNonExistentSession(self):
         self.assertNotIn(b"does-not-exist", request.cookies[0])
 
 
+    def test_getSessionExpired(self):
+        """
+        L{Request.getSession} generates a new session when the previous
+        session has expired.
+        """
+        clock = Clock()
+        site = server.Site(resource.Resource())
+        d = DummyChannel()
+        request = server.Request(d, 1)
+        request.site = site
+        request.sitepath = []
+
+        def SessionFactoryWithClock(site, uid):
+            """
+            Forward to normal session factory, but inject the clock.
+            """
+            session = OriginalSessionFactory(site, uid)
+            session._reactor = clock
+            return session
+
+        # The site is patch to allow injecting a clock to the session.
+        OriginalSessionFactory = site.sessionFactory
+        site.sessionFactory = SessionFactoryWithClock
+
+        initialSession = request.getSession()
+
+        # When the session is requested after the session timeout,
+        # no error is raised and a new session is returned.
+        clock.advance(OriginalSessionFactory.sessionTimeout )
+        newSession = request.getSession()
+
+        self.assertIsNot(initialSession, newSession)
+        self.addCleanup(newSession.expire)
```

Comment:
It's moderately pointless to do `addCleanup` at the _end_ of a test method.  The point is that you do it _before_ any exceptions that could occur and prevent you from doing the cleanup.  The test runner then calls the cleanups you registered whether the test succeeds or fails.  It's basically like a `finally` statement for the whole test method.  You should call `addCleanup` as soon as the session is created. 

But, also, since the IReactorTime is a `Clock` there's probably no actual reason to expire the session.  No resources are being leaked by not expiring it.